### ISSUE TITLE
Break out Blind/Timing Based Injections To Separate Rules

### ISSUE THREAD

**Issue by sgerlach** [zaproxy/zaproxy#7341]
### Is your feature request related to a problem? Please describe.

In at least the SQL Injection (40018) and OS Command Injection (90020) plugins, there are definitive attack/response detections and non-definitive attack/responses usually assigned with blind injection attacks. Those attacks attempt to perform a sleep injection to control the applications timing and perform StdDev work on the responses to see if they were successful. 

The timing based alerts consistently generate False Positives due to a handful of factors, one of them being the application is under heavy load. This makes troubleshooting and diagnosis very hard.

### Describe the solution you'd like

Proposal is to break out those blind attacks to their owns plugins so they can be optimized.

- Allows the rule to be disabled and the other parts of the core rule to run
- Allows the Blind attack to have better retest/validation logic
- Will integrate better with the upcoming retest work to only have to run the blind rules or the non-blind rules

### Describe alternatives you've considered

- Adjustments on timing
- Running ONLY these rules to keep load off the app (does not scale well for automation)
- Adding additional test logic to the current rules to validate alerts

### Would you like to help fix this issue?

- [ ] Yes

## Update - the rules that perform timing based attacks

- [x] [40019](https://www.zaproxy.org/docs/alerts/40019/) (limited unit tests) - MySQL (All payloads seem timing based)
- [x] [40020](https://www.zaproxy.org/docs/alerts/40020/) (limited unit tests) - Hypersonic (All payloads seem timing based)
- [x] [40021](https://www.zaproxy.org/docs/alerts/40021/) (limited unit tests) - Oracle (Union and timing)
- [x] [40022](https://www.zaproxy.org/docs/alerts/40022/) (limited unit tests) - PostgreSQL (All payloads seem timing based)
- [x] [40024](https://www.zaproxy.org/docs/alerts/40024/) (limited unit tests) - SQLite
- [x] [40027](https://www.zaproxy.org/docs/alerts/40027/) (limited unit tests, might only do timing attacks?) - MsSQL
- [x] [40033](https://www.zaproxy.org/docs/alerts/40033/) (limited unit tests) - MongoDB (Split to 90039)
- [ ] [90018](https://www.zaproxy.org/docs/alerts/90018/) (no unit tests) - Adv SQLi
- [x] [90020](https://www.zaproxy.org/docs/alerts/90020/) (limited unit tests) - CMDi (Split to 90037)

Ref for rule IDs: https://github.com/zaproxy/zaproxy/blob/main/docs/scanners.md

**Comment by psiinon** [comment]
I definitely like the idea of people being able to disable all of these checks in one go.
I do worry that moving them all into a new add-on will be difficult, but lets discuss what options we have.
One possibility would be to abuse the 'technology' configs - maybe having a new "Attacks" / "Timing" option which would then work with minimal changes. We might be able to use this for other things?

**Comment by thc202** [comment]
Just splitting into their own scan rules and keeping them in the same add-ons sounds better.

**Comment by psiinon** [comment]
We could tag them as well?

**Comment by thc202** [comment]
Definitely.

**Comment by kingthorin** [comment]
Do you mean with Alert tags? Or do you mean with Tech as you originally suggested?

If you mean Alert tags then yes definitely.

**Comment by psiinon** [comment]
We've discussed this on IRC and we like the idea of splitting the timing checks into different rules, but keeping them in the same add-ons. The alert tags would be a bonus...

**Comment by psiinon** [comment]
I've updated the first comment with a list of all the rules that I could find which perform timing based attacks - anyone think of any that I've missed?

**Comment by DiogoMRSilva** [comment]
In my opinion that shouldn't be  the solution to the time attacks fps. My suggested solutions are: improving the time delay detection logic by making an unique method in the common lib that would deal with the maths and number of requests, don't run time based if confidence is high, use several detection methods (delay, reflection, error) simultaneously and if more than one work increase confidence. 

Some disadvantages of splitting the rules:
- no having different types of detections together in the same rule  to increase confidence. 
- loss of control on the amount of request the user wants to do on each vulnerability, since strength is +- correlated to the number of request if we have both at high  we have twice the number of allowed requests
-  Have many entries to chose in the scan policies

**Comment by sgerlach** [comment]
Thanks for the conversation here. I'm going to bounty each of these tasks to get some 💵 associated with this work.

**Comment by FiveOFive** [comment]
Great to see the PR (https://github.com/zaproxy/zap-extensions/pull/4316) is making it's way through. I'm also seeing a high number of false positives from timing based attacks and looking forward to this change. Let me know if there are outstanding items to get this done that need to be picked up and I can try to help out with them. Thanks!

**Comment by FiveOFive** [comment]
Hi ZAP team - checking in to see if this is still planned?

**Comment by thc202** [comment]
Yes but still pending on fixes being done to the scan rules.

**Comment by FiveOFive** [comment]
Hi [thc202](https://github.com/thc202) anything we can do to help get this through the finish line? I could either try to pick it up or potentially sponsor a member of the ZAP team to work on it.

**Comment by thc202** [comment]
Ticked the ones that were already done, there's also a status in https://github.com/zaproxy/zaproxy/issues/8112#issuecomment-1819848796

**Comment by thc202** [comment]
Which ones are a priority?

**Comment by FiveOFive** [comment]
> Ticked the ones that were already done, there's also a status in [#8112 (comment)](https://github.com/zaproxy/zaproxy/issues/8112#issuecomment-1819848796)

Thanks, I'd missed that some were already done. We'll review and see if any others are still a priority

**Comment by kingthorin** [comment]
The Adv SQLi was originally based on SQLMap, it has diverged greatly. I doubt it will ever be caught up. Does it make sense to break it or shall we leave it?

**Comment by kingthorin** [comment]
https://www.zaproxy.org/blog/2025-07-22-timing-rule-changes/

**Comment by thc202** [comment]
Can't this be closed already along with the other PR?

**Comment by kingthorin** [comment]
Depends what we wanna do with the adv rule.

I think so, but don't want to decide unilaterally.

**Comment by thc202** [comment]
My opinion about that rule is still the same but lets wait.

**Comment by psiinon** [comment]
I'm good with closing it 😁
